sample/sample.go

```diff
@@ -0,0 +1,205 @@
+package sample
+
+import (
+	"cmp"
+	"errors"
+	"math"
+	"slices"
+
+	"golang.org/x/exp/rand"
+	"gonum.org/v1/gonum/floats"
+	"gonum.org/v1/gonum/stat/sampleuv"
+)
+
+type Transform interface {
```

Comment:
What's the rationale for adding this interface?

Reply:
Wanted to make the boundary more clear between logits and choosing an index given logits. So this is `Transform` and then `Sample` not tied to the naming but I think the distinction is important.

Reply:
This seems weird to me. If you want that distinction, maybe `Weighed` and `Greedy` shouldn't implement the `Sampler` interface

e.g.
```go

func Greedy(logits []float32) int {
    ...
}

func Weighed(logits []float32, samplers ...Sampler) int {
    ...
}

func Sample(logits []float32) []int {
    if temperature > 0 {
        return Weighed(
            logits,
            Temperature(temperature),
            TopK(topk),
            ...,
        )
    }

    return Greedy(logits)
}
```

Reply:
I was thinking maybe something along the lines of 
```golang
type Transform interface {
	Apply([]float64) ([]float64, error)
}

type Sampler interface {
	Sample([]float32, ...Transform) (int, error)
}
```
it might be a bit extra for what we need but does make the inputs outputs super clear for the samplers